reducer/src/main/java/com/graphicsfuzz/reducer/glslreducers/EliminateInjectionMacrosVisitor.java

```diff
@@ -16,128 +16,75 @@
+          || MacroNames.isFalse(functionCallExpr)
+          || MacroNames.isTrue(functionCallExpr)) {
+        parent.replaceChild(child,
+            addParenthesesIfNecessary(parent, functionCallExpr.getChild(1)));
+      } else if (MacroNames.isFuzzed(functionCallExpr)
+          || MacroNames.isDeadByConstruction(functionCallExpr)
+          || MacroNames.isSwitch(functionCallExpr)
+          || MacroNames.isLoopWrapper(functionCallExpr)
+          || MacroNames.isIfWrapperFalse(functionCallExpr)
+          || MacroNames.isIfWrapperTrue(functionCallExpr)
+      ) {
+        parent.replaceChild(child,
+            addParenthesesIfNecessary(parent, functionCallExpr.getChild(0)));
+      }
     }
   }
 
-  private void cleanupLoop(LoopStmt loopStmt) {
-    if (MacroNames.isLoopWrapper(loopStmt.getCondition())) {
-      loopStmt.setCondition(loopStmt.getCondition().getChild(0));
+  private IAstNode addParenthesesIfNecessary(IAstNode parent, Expr child) {
+    if (child instanceof ConstantExpr
+        || child instanceof ParenExpr
+        || child instanceof VariableIdentifierExpr
+        || child instanceof FunctionCallExpr) {
+      // Parentheses is unnecessary in cases such as _GLF_FUNCTION(1),
+      // _GLF_FUNCTION((1)), _GLF_FUNCTION(a), _GLF_FUNCTION(sin(a)).
+      return child;
     }
-  }
-
-  @Override
-  public void visitDoStmt(DoStmt doStmt) {
-    super.visitDoStmt(doStmt);
-    cleanupLoop(doStmt);
-  }
 
-  @Override
-  public void visitForStmt(ForStmt forStmt) {
-    super.visitForStmt(forStmt);
-    cleanupLoop(forStmt);
-  }
+    if (!(parent instanceof Expr)) {
+      // No parentheses needed if the parent is not an expression,
+      // for example, int x = _GLF_FUNCTION(a + b).
+      return child;
+    }
 
-  private void cleanUpMacros(Expr parent) {
-    for (int i = 0; i < parent.getNumChildren(); i++) {
-      Expr child = parent.getChild(i);
-      // Note: it would be redundant to have a special function reduction opportunity
-      // be classed also as a fuzzed expression reduction opportunity
-      if (MacroNames.isIdentity(child)
-          || MacroNames.isZero(child)
-          || MacroNames.isOne(child)
-          || MacroNames.isFalse(child)
-          || MacroNames.isTrue(child)) {
-        replaceChildWithGrandchild(parent, i, 1);
-      } else if (MacroNames.isFuzzed(child)
-          || MacroNames.isDeadByConstruction(child)) {
-        replaceChildWithGrandchild(parent, i, 0);
-      }
+    if (parent instanceof ParenExpr) {
+      // If parent is parentheses, adding a new parentheses would be redundant,
+      // e.g. (_GLF_FUNCTION(a + b)).
+      return child;
     }
-  }
 
-  @Override
-  public void visitScalarInitializer(ScalarInitializer scalarInitializer) {
-    super.visitScalarInitializer(scalarInitializer);
-    if (MacroNames.isFuzzed(scalarInitializer.getExpr())) {
-      scalarInitializer.setExpr(scalarInitializer.getExpr().getChild(0));
+    if (parent instanceof FunctionCallExpr) {
```

Comment:
For completeness, could you add something like:

      // If parent is a function call, adding a new parentheses would be redundant,
      // e.g. foo(..., _GLF_FUNCTION(a + b), ...).

?

I really like the other comments you have added like this, so it would be great to do the same in this case too.

Reply:
Done!